### PR TITLE
fix(clustering): hybrid mode should accept special characters used in forward proxy password

### PR DESCRIPTION
**Pull request by ms2008**
### Summary

Hybrid mode connections not working if the forward proxy password contains special character(`#`). Even after the `proxy_server` is url encoded, it still doesn't work and reports a "407 Proxy Authentication Required" error.

This fix tries to address this issue. But, keep in mind that even with this fix, `proxy_server` still needs to be url-encoded.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix FTI-6145

### CODE REVIEW

File: changelog/unreleased/kong/fix-clustering-forward-proxy-authentication.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+message: Fixed an issue where hybrid mode not working if the forward proxy password contains special character(#). Note that the `proxy_server` configuration parameter still needs to be url-encoded.
```
[the remark this answers is not in the document]
Comment (ms2008) [reply changelog/unreleased/kong/fix-clustering-forward-proxy-authentication.yml:1]: It's feasible, but the implementation may not be so straightforward. Because if we do this way, there's a caveat: it requires that the user's input is not URL-encoded. If the user's input is already URL-encoded, then it would be encoded twice here, which obviously is incorrect.

To avoid this situation, we need to decode the URL once before parsing it using `url_parse`, ensuring we get plain text. Afterward, we can encode it again for `url_parse` to parse. Personally, I think the process is a bit wordy.

Additionally, `curl` expects users to handle URL encoding themselves in such cases as well:

```
$ curl -x 'http://testuser:rye#!ihef4@localhost:3128' http://httpbin-1/stream/5
curl: (5) Unsupported proxy syntax in 'http://testuser:rye#!ihef4@localhost:3128'
```

[the remark this answers is not in the document]
Comment (liverpool8056) [reply changelog/unreleased/kong/fix-clustering-forward-proxy-authentication.yml:1]: Make sense. This requires users to provide an encoded string, so should we change the comment in the `kong.conf.default` a bit:

```
#proxy_server =                  # Proxy server defined as an encoded URL. Kong will only use this
```

**Review by liverpool8056 — APPROVED** [review]

File: changelog/unreleased/kong/fix-clustering-forward-proxy-authentication.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ms2008) [reply changelog/unreleased/kong/fix-clustering-forward-proxy-authentication.yml:1]: done

File: kong/clustering/utils.lua [hunk 2]
```diff
@@ -50,7 +51,7 @@ local function parse_proxy_url(proxy_server)
     end
 
     if parsed.user and parsed.password then
-      ret.proxy_authorization = "Basic " .. encode_base64(parsed.user  .. ":" .. parsed.password)
+      ret.proxy_authorization = "Basic " .. encode_base64(unescape_uri(parsed.user)  .. ":" .. unescape_uri(parsed.password))
```
Comment (catbro666) [inline kong/clustering/utils.lua:54]: Should we also unescape the other parts of the url, for instance, the host?

Comment (ms2008) [reply kong/clustering/utils.lua:54]: I've actually flashed this issue in my head before, and I'm guessing there aren't many cases where the `host` field uses special characters. But just in case, I'll deal with the situation now.

**Review by catbro666 — APPROVED** [review]
